src/slurm_plugin/cluster_event_publisher.py

```diff
@@ -0,0 +1,371 @@
+                        "count": len(invalid_backing_instance_nodes),
+                        "nodes": self._generate_node_name_list(invalid_backing_instance_nodes),
+                    }
+                }
+
+        timestamp = ClusterEventPublisher.timestamp()
+
+        self.publish_event(
+            "WARNING",
+            "Number of nodes without a valid backing instance",
+            "invalid-backing-instance-count",
+            timestamp=timestamp,
+            event_supplier=detail_supplier(),
+        )
+
+    @log_exception(logger, "publish_bootstrap_failure_events", catch_exception=Exception, raise_on_error=False)
+    def publish_bootstrap_failure_events(self, bootstrap_failure_nodes: List[SlurmNode]):
+        def dynamic_node_filter(node: SlurmNode) -> bool:
+            return isinstance(node, DynamicNode) and node.is_resume_failed() and node.is_nodeaddr_set()
+
+        def static_node_filter(node: SlurmNode) -> bool:
+            return isinstance(node, StaticNode) and node._is_replacement_timeout
+
+        def protected_node_list(filter: Callable[[SlurmNode], bool]):
+            return [node for node in bootstrap_failure_nodes if filter(node)]
+
+        def protected_mode_error_count_supplier():
+            dynamic_nodes = protected_node_list(dynamic_node_filter)
+            static_nodes = protected_node_list(static_node_filter)
+            if dynamic_nodes or static_nodes:
+                yield {
+                    "detail": {
+                        "count": len(dynamic_nodes) + len(static_nodes),
+                        "static-replacement-timeout-errors": {
+                            "count": len(static_nodes),
+                            "nodes": self._generate_node_name_list(static_nodes),
+                        },
+                        "dynamic-resume-timeout-errors": {
+                            "count": len(static_nodes),
+                            "nodes": self._generate_node_name_list(dynamic_nodes),
+                        },
+                    }
+                }
+
+        def bootstrap_failure_supplier():
+            if bootstrap_failure_nodes:
+                yield {
+                    "detail": {
+                        "count": len(bootstrap_failure_nodes),
+                        "nodes": self._generate_node_name_list(bootstrap_failure_nodes),
+                    }
+                }
+
+        timestamp = ClusterEventPublisher.timestamp()
+
+        self.publish_event(
+            "WARNING",
+            "Number of nodes that failed to bootstrap",
+            "protected-mode-error-count",
+            timestamp=timestamp,
+            event_supplier=protected_mode_error_count_supplier(),
+        )
+
+        self.publish_event(
+            "DEBUG",
+            "Number of nodes that failed to bootstrap",
+            "bootstrap-failure-count",
+            timestamp=timestamp,
+            event_supplier=bootstrap_failure_supplier(),
+        )
```

Comment:
Instead of publishing two events here,
what about we only publish one warning event, that have "static-replacement-timeout-errors", "dynamic-resume-timeout-errors" and "other-bootstrap-errors" in `detail`?

Reply:
Changed. There is now an "other" for other errors that are not bootstrap timeouts.